In [7]:
#Import Libraries
import pandas as pd
import numpy as np

In [8]:
#Load dataset
df = pd.read_csv("../data/raw/train.csv")

C:\Users\a\AppData\Local\Temp\ipykernel_21152\4286747745.py:2: DtypeWarning: Columns (26) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("../data/raw/train.csv")


In [9]:
#Always preserve the original dataset and create a copy of it to work with.
df_clean = df.copy()

In [10]:
df_clean.shape

(100000, 28)

In [11]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 28 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   ID                        100000 non-null  object 
 1   Customer_ID               100000 non-null  object 
 2   Month                     100000 non-null  object 
 3   Name                      90015 non-null   object 
 4   Age                       100000 non-null  object 
 5   SSN                       100000 non-null  object 
 6   Occupation                100000 non-null  object 
 7   Annual_Income             100000 non-null  object 
 8   Monthly_Inhand_Salary     84998 non-null   float64
 9   Num_Bank_Accounts         100000 non-null  int64  
 10  Num_Credit_Card           100000 non-null  int64  
 11  Interest_Rate             100000 non-null  int64  
 12  Num_of_Loan               100000 non-null  object 
 13  Type_of_Loan              88592 non-null   ob

In [12]:
df_clean.head()

,ID,Customer_ID,Month,Name,Age,SSN,Occupation,Annual_Income,Monthly_Inhand_Salary,Num_Bank_Accounts,...,Credit_Mix,Outstanding_Debt,Credit_Utilization_Ratio,Credit_History_Age,Payment_of_Min_Amount,Total_EMI_per_month,Amount_invested_monthly,Payment_Behaviour,Monthly_Balance,Credit_Score
0,0x1602,CUS_0xd40,January,Aaron Maashoh,23,821-00-0265,Scientist,19114.12,1824.843333,3,...,_,809.98,26.822620,22 Years and 1 Months,No,49.574949,80.41529543900253,High_spent_Small_value_payments,312.49408867943663,Good
1,0x1603,CUS_0xd40,February,Aaron Maashoh,23,821-00-0265,Scientist,19114.12,NaN,3,...,Good,809.98,31.944960,NaN,No,49.574949,118.28022162236736,Low_spent_Large_value_payments,284.62916249607184,Good
2,0x1604,CUS_0xd40,March,Aaron Maashoh,-500,821-00-0265,Scientist,19114.12,NaN,3,...,Good,809.98,28.609352,22 Years and 3 Months,No,49.574949,81.699521264648,Low_spent_Medium_value_payments,331.2098628537912,Good
3,0x1605,CUS_0xd40,April,Aaron Maashoh,23,821-00-0265,Scientist,19114.12,NaN,3,...,Good,809.98,31.377862,22 Years and 4 Months,No,49.574949,199.4580743910713,Low_spent_Small_value_payments,223.45130972736786,Good
4,0x1606,CUS_0xd40,May,Aaron Maashoh,23,821-00-0265,Scientist,19114.12,1824.843333,3,...,Good,809.98,24.797347,22 Years and 5 Months,No,49.574949,41.420153086217326,High_spent_Medium_value_payments,341.48923103222177,Good


In [13]:
df_clean.columns

Index(['ID', 'Customer_ID', 'Month', 'Name', 'Age', 'SSN', 'Occupation',
       'Annual_Income', 'Monthly_Inhand_Salary', 'Num_Bank_Accounts',
       'Num_Credit_Card', 'Interest_Rate', 'Num_of_Loan', 'Type_of_Loan',
       'Delay_from_due_date', 'Num_of_Delayed_Payment', 'Changed_Credit_Limit',
       'Num_Credit_Inquiries', 'Credit_Mix', 'Outstanding_Debt',
       'Credit_Utilization_Ratio', 'Credit_History_Age',
       'Payment_of_Min_Amount', 'Total_EMI_per_month',
       'Amount_invested_monthly', 'Payment_Behaviour', 'Monthly_Balance',
       'Credit_Score'],
      dtype='object')

In [14]:
#Remove unnecessary columns like ID, Customer_ID, Name, SSN bcz these don't help predict credit score and may contain the sensitive information.
df_clean.drop(['ID', 'Customer_ID', 'Name', 'SSN'], axis=1, inplace=True)


In [15]:
#Detect missing values
df_clean.isnull().sum()

Month                           0
Age                             0
Occupation                      0
Annual_Income                   0
Monthly_Inhand_Salary       15002
Num_Bank_Accounts               0
Num_Credit_Card                 0
Interest_Rate                   0
Num_of_Loan                     0
Type_of_Loan                11408
Delay_from_due_date             0
Num_of_Delayed_Payment       7002
Changed_Credit_Limit            0
Num_Credit_Inquiries         1965
Credit_Mix                      0
Outstanding_Debt                0
Credit_Utilization_Ratio        0
Credit_History_Age           9030
Payment_of_Min_Amount           0
Total_EMI_per_month             0
Amount_invested_monthly      4479
Payment_Behaviour               0
Monthly_Balance              1200
Credit_Score                    0
dtype: int64

In [16]:
#Missing%
(df_clean.isnull().mean()*100).sort_values(ascending=False)

Monthly_Inhand_Salary       15.002
Type_of_Loan                11.408
Credit_History_Age           9.030
Num_of_Delayed_Payment       7.002
Amount_invested_monthly      4.479
Num_Credit_Inquiries         1.965
Monthly_Balance              1.200
Month                        0.000
Credit_Mix                   0.000
Payment_Behaviour            0.000
Total_EMI_per_month          0.000
Payment_of_Min_Amount        0.000
Credit_Utilization_Ratio     0.000
Outstanding_Debt             0.000
Changed_Credit_Limit         0.000
Age                          0.000
Delay_from_due_date          0.000
Num_of_Loan                  0.000
Interest_Rate                0.000
Num_Credit_Card              0.000
Num_Bank_Accounts            0.000
Annual_Income                0.000
Occupation                   0.000
Credit_Score                 0.000
dtype: float64

Handling missing values in numerical columns by using median. 
Handling missing values in categorical columns using mode

In [17]:
#Check duplicate records
df_clean.duplicated().sum()

np.int64(0)

In [18]:
#Remove duplicate rows but in this dataset,no duplicates
df_clean.drop_duplicates(inplace=True)

In [19]:
#Check data types
df_clean.dtypes

Month                        object
Age                          object
Occupation                   object
Annual_Income                object
Monthly_Inhand_Salary       float64
Num_Bank_Accounts             int64
Num_Credit_Card               int64
Interest_Rate                 int64
Num_of_Loan                  object
Type_of_Loan                 object
Delay_from_due_date           int64
Num_of_Delayed_Payment       object
Changed_Credit_Limit         object
Num_Credit_Inquiries        float64
Credit_Mix                   object
Outstanding_Debt             object
Credit_Utilization_Ratio    float64
Credit_History_Age           object
Payment_of_Min_Amount        object
Total_EMI_per_month         float64
Amount_invested_monthly      object
Payment_Behaviour            object
Monthly_Balance              object
Credit_Score                 object
dtype: object

Many columns in this dataset are stored as strings because of symbols or invalid values. Convert numeric columns where needed. Some columns contain commas, underscores, or other symbols.

AGE

In [20]:
df_clean["Age"].head(20)

0       23
1       23
2     -500
3       23
4       23
5       23
6       23
7       23
8      28_
9       28
10      28
11      28
12      28
13      28
14      28
15      28
16      34
17      34
18      34
19      34
Name: Age, dtype: object

In [21]:
df_clean["Age"].unique()[:20]  #unique values

array(['23', '-500', '28_', '28', '34', '54', '55', '21', '31', '33',
       '34_', '7580', '30', '30_', '24', '24_', '44', '45', '40', '41'],
      dtype=object)

Sometimes negative or unrealistic values exist.

In [22]:
#convert into numeric
df_clean["Age"] = pd.to_numeric(df_clean["Age"], errors="coerce")

In [23]:
#Check range
df_clean["Age"].describe()

count    95061.000000
mean       110.934505
std        689.407864
min       -500.000000
25%         24.000000
50%         33.000000
75%         42.000000
max       8698.000000
Name: Age, dtype: float64

In [24]:
#find and handle impossible values (unrealistic ages)
df_clean[(df_clean["Age"]<18) | (df_clean["Age"]>100)]

,Month,Age,Occupation,Annual_Income,Monthly_Inhand_Salary,Num_Bank_Accounts,Num_Credit_Card,Interest_Rate,Num_of_Loan,Type_of_Loan,...,Credit_Mix,Outstanding_Debt,Credit_Utilization_Ratio,Credit_History_Age,Payment_of_Min_Amount,Total_EMI_per_month,Amount_invested_monthly,Payment_Behaviour,Monthly_Balance,Credit_Score
2,March,-500.0,Scientist,19114.12,NaN,3,4,3,4,"Auto Loan, Credit-Builder Loan, Personal Loan,...",...,Good,809.98,28.609352,22 Years and 3 Months,No,49.574949,81.699521264648,Low_spent_Medium_value_payments,331.2098628537912,Good
56,January,7580.0,Media_Manager,34081.38_,NaN,8,7,15,3,"Not Specified, Auto Loan, and Student Loan",...,Standard,1704.18,24.448063,NaN,NM,70.478333,162.4410091967751,Low_spent_Large_value_payments,298.19215813115227,Poor
113,February,-500.0,Entrepreneur,25546.26,NaN,8,7,14,5,"Not Specified, Student Loan, Student Loan, Cre...",...,Standard,758.44,29.711376,18 Years and 3 Months,Yes,101.328637,300.32323168069064,Low_spent_Small_value_payments,129.9336309554582,Standard
122,March,181.0,Scientist,31993.78,2942.148333,6,6,7,2,"Payday Loan, and Home Equity Loan",...,Standard,818.22,27.380109,17 Years and 0 Months,Yes,45.141298,264.25708850339777,Low_spent_Small_value_payments,274.8164466052998,Standard
219,April,995.0,Accountant,43070.24,3622.186667,3,3,18,1,Debt Consolidation Loan,...,Standard,1233.1,24.331772,19 Years and 5 Months,Yes,30.576085,74.92037487328841,High_spent_Medium_value_payments,506.7222069702167,Standard
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
99913,February,2263.0,Media_Manager,20059.98,1523.665000,8,5,12,4,"Home Equity Loan, Payday Loan, Not Specified, ...",...,Good,909.01,25.982154,16 Years and 4 Months,No,45.076827,74.71358019803843,High_spent_Small_value_payments,292.576093,Poor
99937,February,-500.0,_______,22620.79,1722.065833,7,3,9,0_,NaN,...,Standard,642.46,31.841872,NaN,No,0.000000,105.07629339039072,Low_spent_Large_value_payments,337.13029,Standard
99950,July,1342.0,Media_Manager,59146.36,4908.863333,3,6,6,1,Personal Loan,...,_,418.03,38.199635,20 Years and 7 Months,No,26.778419,502.3763201756033,Low_spent_Small_value_payments,251.731594,Standard
99963,April,-500.0,Lawyer,42903.79,NaN,0,4,6,1,Not Specified,...,Good,1079.48,30.625298,NaN,No,34.975457,31.19391926667726,High_spent_Large_value_payments,520.662207,Standard


In [25]:
#Replace with NaN
df_clean.loc[
(df_clean["Age"]<18) |
(df_clean["Age"]>100),
"Age"] = np.nan

In [26]:
df_clean["Age"].fillna(df_clean["Age"].median(), inplace=True)

C:\Users\a\AppData\Local\Temp\ipykernel_21152\3691372764.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_clean["Age"].fillna(df_clean["Age"].median(), inplace=True)


In [27]:
df_clean["Age"].head(20)

0     23.0
1     23.0
2     34.0
3     23.0
4     23.0
5     23.0
6     23.0
7     23.0
8     34.0
9     28.0
10    28.0
11    28.0
12    28.0
13    28.0
14    28.0
15    28.0
16    34.0
17    34.0
18    34.0
19    34.0
Name: Age, dtype: float64

ANNUAL_INCOME

In [28]:
df_clean["Annual_Income"].head(20)

0      19114.12
1      19114.12
2      19114.12
3      19114.12
4      19114.12
5      19114.12
6      19114.12
7      19114.12
8      34847.84
9      34847.84
10    34847.84_
11     34847.84
12     34847.84
13     34847.84
14     34847.84
15     34847.84
16    143162.64
17    143162.64
18    143162.64
19    143162.64
Name: Annual_Income, dtype: object

In [29]:
#Remove underscore and convert into numeric 
df_clean["Annual_Income"] = (
    df_clean["Annual_Income"]
    .astype(str)
    .str.replace("_", "",regex=False)
)
df_clean["Annual_Income"] = pd.to_numeric(df_clean["Annual_Income"], errors="coerce")

In [30]:
#Handle missing values in Annual_Income by replacing them with the median value
df_clean["Annual_Income"].fillna(
df_clean["Annual_Income"].median(),
inplace=True)

C:\Users\a\AppData\Local\Temp\ipykernel_21152\3031593351.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_clean["Annual_Income"].fillna(


In [31]:
df_clean["Annual_Income"].head(20)

0      19114.12
1      19114.12
2      19114.12
3      19114.12
4      19114.12
5      19114.12
6      19114.12
7      19114.12
8      34847.84
9      34847.84
10     34847.84
11     34847.84
12     34847.84
13     34847.84
14     34847.84
15     34847.84
16    143162.64
17    143162.64
18    143162.64
19    143162.64
Name: Annual_Income, dtype: float64

Monthly_Inhand_Salary

In [32]:
df_clean["Monthly_Inhand_Salary"].head(20)

0      1824.843333
1              NaN
2              NaN
3              NaN
4      1824.843333
5              NaN
6      1824.843333
7      1824.843333
8      3037.986667
9      3037.986667
10     3037.986667
11             NaN
12     3037.986667
13     3037.986667
14             NaN
15     3037.986667
16    12187.220000
17    12187.220000
18             NaN
19    12187.220000
Name: Monthly_Inhand_Salary, dtype: float64

In [33]:
df_clean["Monthly_Inhand_Salary"] = pd.to_numeric(
df_clean["Monthly_Inhand_Salary"],
errors="coerce")

In [34]:
df_clean["Monthly_Inhand_Salary"].fillna(
df_clean["Monthly_Inhand_Salary"].median(),
inplace=True)

C:\Users\a\AppData\Local\Temp\ipykernel_21152\1759563927.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_clean["Monthly_Inhand_Salary"].fillna(


In [35]:
df_clean["Monthly_Inhand_Salary"].head(20)

0      1824.843333
1      3093.745000
2      3093.745000
3      3093.745000
4      1824.843333
5      3093.745000
6      1824.843333
7      1824.843333
8      3037.986667
9      3037.986667
10     3037.986667
11     3093.745000
12     3037.986667
13     3037.986667
14     3093.745000
15     3037.986667
16    12187.220000
17    12187.220000
18     3093.745000
19    12187.220000
Name: Monthly_Inhand_Salary, dtype: float64

Num_Bank_Accounts

In [36]:
df_clean["Num_Bank_Accounts"].head(20)

0     3
1     3
2     3
3     3
4     3
5     3
6     3
7     3
8     2
9     2
10    2
11    2
12    2
13    2
14    2
15    2
16    1
17    1
18    1
19    1
Name: Num_Bank_Accounts, dtype: int64

In [37]:
df_clean["Num_Bank_Accounts"] = pd.to_numeric(
df_clean["Num_Bank_Accounts"],
errors="coerce")

In [38]:
df_clean["Num_Bank_Accounts"].describe()

count    100000.000000
mean         17.091280
std         117.404834
min          -1.000000
25%           3.000000
50%           6.000000
75%           7.000000
max        1798.000000
Name: Num_Bank_Accounts, dtype: float64

In [39]:
df_clean["Num_Bank_Accounts"].value_counts().sort_index()

Num_Bank_Accounts
-1          21
 0        4328
 1        4490
 2        4304
 3       11950
         ...  
 1786        1
 1789        2
 1793        1
 1794        2
 1798        2
Name: count, Length: 943, dtype: int64

In [40]:
df_clean[df_clean["Num_Bank_Accounts"] > 20]

,Month,Age,Occupation,Annual_Income,Monthly_Inhand_Salary,Num_Bank_Accounts,Num_Credit_Card,Interest_Rate,Num_of_Loan,Type_of_Loan,...,Credit_Mix,Outstanding_Debt,Credit_Utilization_Ratio,Credit_History_Age,Payment_of_Min_Amount,Total_EMI_per_month,Amount_invested_monthly,Payment_Behaviour,Monthly_Balance,Credit_Score
267,April,44.0,Writer,58317.00,3093.745000,1414,5,9,6,"Mortgage Loan, Personal Loan, Home Equity Loan...",...,Standard,98.97,29.766107,NaN,Yes,149.897199,158.64827595455043,High_spent_Medium_value_payments,407.9295246244412,Standard
288,January,34.0,Mechanic,29469.98,2227.831667,1231,7,24,5,"Personal Loan, Credit-Builder Loan, Debt Conso...",...,_,3421.66,24.639658,NaN,Yes,69.685459,59.825596124675045,!@9#%8,363.27211196971183,Standard
310,July,37.0,Musician,15566.02,1423.168333,67,5,32,4,"Home Equity Loan, Credit-Builder Loan, Persona...",...,Standard,1693.95,29.706454,8 Years and 4 Months,Yes,43.070520,80.48442010154183,!@9#%8,308.76189362947565,Poor
339,April,42.0,_______,20574.47,3093.745000,572,3,14,4,"Credit-Builder Loan, Personal Loan, Credit-Bui...",...,Standard,749.95,36.559538,11 Years and 2 Months,Yes,49.348666,25.161404432750263,High_spent_Medium_value_payments,349.54384591241774,Standard
356,May,19.0,Musician,85554.03,7185.502500,1488,2,3,-100,NaN,...,_,1095.73,41.661802,19 Years and 11 Months,No,0.000000,70.82263261910934,High_spent_Large_value_payments,887.7276173808908,Good
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
99591,August,29.0,Mechanic,17013.29,1326.774167,813,1,10,1,Debt Consolidation Loan,...,_,1452.79,28.051684,32 Years and 6 Months,NM,13.109663,55.72695329443804,Low_spent_Small_value_payments,353.840801,Good
99638,July,25.0,Developer,125271.96,10374.330000,1481,7,1,4,"Mortgage Loan, Debt Consolidation Loan, Mortga...",...,Good,827.56,33.201730,25 Years and 8 Months,NM,241.065885,180.56001463373025,!@9#%8,855.8071,Poor
99666,March,31.0,Lawyer,146310.68,12124.556667,474,4,9,1,Debt Consolidation Loan,...,Good,928.28,43.274889,22 Years and 3 Months,No,72.250125,121.28482498252883,High_spent_Large_value_payments,1258.920717,Standard
99722,March,53.0,Architect,36817.94,3198.161667,697,4,5,3,"Home Equity Loan, Auto Loan, and Mortgage Loan",...,Good,1019.46,26.578799,16 Years and 9 Months,No,86.809918,123.91555910514948,Low_spent_Small_value_payments,399.09069,Standard


In [41]:
df_clean["Num_Bank_Accounts"].value_counts().sort_index().loc[15:30]

Num_Bank_Accounts
18    1
26    1
27    1
28    3
29    1
30    1
Name: count, dtype: int64

In [42]:
#handle negative values
df_clean.loc[df_clean["Num_Bank_Accounts"] < 0, "Num_Bank_Accounts"] = np.nan
#treat unrealistic values as missing
df_clean.loc[df_clean["Num_Bank_Accounts"] > 30, "Num_Bank_Accounts"] = np.nan
#fill
df_clean["Num_Bank_Accounts"] = df_clean["Num_Bank_Accounts"].fillna(
    df_clean["Num_Bank_Accounts"].median()
)

In [43]:
df_clean["Num_Bank_Accounts"].describe()

count    100000.000000
mean          5.365790
std           2.582703
min           0.000000
25%           3.000000
50%           5.000000
75%           7.000000
max          30.000000
Name: Num_Bank_Accounts, dtype: float64

The Num_Bank_Accounts column contained invalid negative values and extremely large values (up to 1798), which are unrealistic for individual customers. Inspection of the distribution showed that values between 18 and 30 were very rare but present, while values above 30 were anomalous. Therefore, values less than 0 and greater than 30 were treated as invalid, replaced with missing values, and imputed using the median.

Num_Credit_Card

In [44]:
df_clean["Num_Credit_Card"].head(20)

0        4
1        4
2        4
3        4
4        4
5        4
6        4
7        4
8        4
9        4
10    1385
11       4
12       4
13       4
14       4
15       4
16       5
17       5
18       5
19       5
Name: Num_Credit_Card, dtype: int64

In [45]:
df_clean["Num_Credit_Card"] = pd.to_numeric(
df_clean["Num_Credit_Card"],
errors="coerce")

In [46]:
df_clean["Num_Credit_Card"].describe()

count    100000.00000
mean         22.47443
std         129.05741
min           0.00000
25%           4.00000
50%           5.00000
75%           7.00000
max        1499.00000
Name: Num_Credit_Card, dtype: float64

In [47]:
df_clean["Num_Credit_Card"].value_counts().sort_index().loc[10:30]

Num_Credit_Card
10    4860
11      36
15       3
16       2
17       1
18       1
20       1
21       1
22       2
24       1
25       5
26       1
27       3
28       4
29       2
Name: count, dtype: int64

In [48]:
df_clean[df_clean["Num_Credit_Card"] > 30].shape

(2244, 24)

In [49]:
df_clean[df_clean["Num_Credit_Card"] > 30]["Num_Credit_Card"].describe()

count    2244.000000
mean      760.273619
std       430.441073
min        31.000000
25%       373.000000
50%       764.000000
75%      1132.250000
max      1499.000000
Name: Num_Credit_Card, dtype: float64

In [50]:
df_clean[df_clean["Num_Credit_Card"] > 30]["Num_Credit_Card"].value_counts().sort_index().head(20)

Num_Credit_Card
31    2
33    2
34    2
35    1
36    3
37    4
38    1
39    3
41    1
43    4
44    2
46    1
48    2
49    1
50    1
51    4
52    2
53    1
54    2
56    1
Name: count, dtype: int64

In [51]:
# Replace unrealistic values
df_clean.loc[df_clean["Num_Credit_Card"] > 30, "Num_Credit_Card"] = np.nan

# Fill with median
df_clean["Num_Credit_Card"] = df_clean["Num_Credit_Card"].fillna(
    df_clean["Num_Credit_Card"].median()
)

In [52]:
df_clean["Num_Credit_Card"].describe()

count    100000.000000
mean          5.526090
std           2.067332
min           0.000000
25%           4.000000
50%           5.000000
75%           7.000000
max          29.000000
Name: Num_Credit_Card, dtype: float64

In [53]:
df_clean["Num_Credit_Card"].value_counts().sort_index().tail(10)

Num_Credit_Card
18.0    1
20.0    1
21.0    1
22.0    2
24.0    1
25.0    5
26.0    1
27.0    3
28.0    4
29.0    2
Name: count, dtype: int64

The Num_Credit_Card feature contained extreme values ranging up to 1499. Distribution analysis showed that only a few observations existed between 31 and 60, whereas the majority of records above 30 were in the hundreds (median = 764). This strongly indicates that these are corrupted values intentionally inserted into the Kaggle dataset. Since such values are unrealistic for an individual customer, all values greater than 30 were treated as invalid, replaced with missing values, and imputed using the median value.

Interest_Rate

In [54]:
df_clean["Interest_Rate"].head(20)

0     3
1     3
2     3
3     3
4     3
5     3
6     3
7     3
8     6
9     6
10    6
11    6
12    6
13    6
14    6
15    6
16    8
17    8
18    8
19    8
Name: Interest_Rate, dtype: int64

In [55]:
df_clean["Interest_Rate"].describe()

count    100000.000000
mean         72.466040
std         466.422621
min           1.000000
25%           8.000000
50%          13.000000
75%          20.000000
max        5797.000000
Name: Interest_Rate, dtype: float64

In [56]:
df_clean["Interest_Rate"] = pd.to_numeric(
df_clean["Interest_Rate"],
errors="coerce")

In [57]:
df_clean[df_clean["Interest_Rate"]>100]

,Month,Age,Occupation,Annual_Income,Monthly_Inhand_Salary,Num_Bank_Accounts,Num_Credit_Card,Interest_Rate,Num_of_Loan,Type_of_Loan,...,Credit_Mix,Outstanding_Debt,Credit_Utilization_Ratio,Credit_History_Age,Payment_of_Min_Amount,Total_EMI_per_month,Amount_invested_monthly,Payment_Behaviour,Monthly_Balance,Credit_Score
44,May,31.0,Lawyer,73928.46,5988.705000,4.0,5.0,5318,0,NaN,...,_,548.2,26.333310,32 Years and 3 Months,No,0.000000,177.95183568608738,Low_spent_Large_value_payments,690.9186643139128,Good
89,February,34.0,Engineer,88640.24,7266.686667,3.0,6.0,433,2,"Payday Loan, and Payday Loan",...,Good,950.36,41.036168,25 Years and 6 Months,NM,135.173371,157.174788304235,High_spent_Large_value_payments,674.3205073907143,Good
167,August,41.0,Musician,19514.88,3093.745000,7.0,5.0,5240,0,NaN,...,_,881.4,28.781888,19 Years and 10 Months,No,0.000000,99.1415414879628,High_spent_Small_value_payments,343.6824585120372,Poor
178,March,20.0,Entrepreneur,106733.13,8873.427500,4.0,4.0,4975,0,NaN,...,Good,76.23,42.070459,32 Years and 11 Months,No,0.000000,172.03282853259896,High_spent_Large_value_payments,955.3099214674013,Good
229,June,40.0,Architect,28572.39,3093.745000,6.0,3.0,668,3,"Payday Loan, Not Specified, and Auto Loan",...,Standard,400.07,35.693949,30 Years and 8 Months,Yes,69.602920,129.264834657262,Low_spent_Small_value_payments,312.13549486074794,Standard
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
99791,August,54.0,Musician,99520.50,8479.375000,3.0,1.0,4396,3,"Home Equity Loan, Not Specified, and Debt Cons...",...,Good,547.21,38.798114,16 Years and 8 Months,NM,196.528591,147.56351637381565,!@9#%8,743.845393,Good
99801,February,34.0,Developer,14937.49,1142.790833,8.0,8.0,387,-100,"Student Loan, Mortgage Loan, Debt Consolidatio...",...,Bad,3699.44,35.964828,13 Years and 3 Months,Yes,50.960878,149.8878636710056,Low_spent_Small_value_payments,203.430341,Standard
99882,March,31.0,Mechanic,16884797.00,5440.945000,7.0,3.0,1947,0,NaN,...,Standard,330.6,38.524095,23 Years and 0 Months,Yes,0.000000,__10000__,High_spent_Small_value_payments,536.34292,Standard
99914,March,54.0,Media_Manager,20059.98,1523.665000,8.0,5.0,360,4,"Home Equity Loan, Payday Loan, Not Specified, ...",...,Good,909.01,29.204711,16 Years and 5 Months,No,45.076827,111.56713873012444,Low_spent_Large_value_payments,265.722534,Poor


Num_of_Loan

In [58]:
df_clean["Num_of_Loan"].tail(20)

99980       3
99981       3
99982       3
99983       3
99984      5_
99985       5
99986       5
99987    -100
99988       5
99989       5
99990       5
99991       5
99992       2
99993       2
99994       2
99995       2
99996       2
99997       2
99998       2
99999       2
Name: Num_of_Loan, dtype: object

In [59]:
df_clean["Num_of_Loan"].describe()

count     100000
unique       434
top            3
freq       14386
Name: Num_of_Loan, dtype: object

In [60]:
df_clean["Num_of_Loan"] = (
df_clean["Num_of_Loan"]
.astype(str)
.str.replace("_","",regex=False)
)

In [61]:
df_clean["Num_of_Loan"] = pd.to_numeric(df_clean["Num_of_Loan"], errors="coerce")
df_clean.loc[
    df_clean["Num_of_Loan"] < 0,
    "Num_of_Loan"
] = np.nan

In [62]:
df_clean["Num_of_Loan"].tail(20)

99980    3.0
99981    3.0
99982    3.0
99983    3.0
99984    5.0
99985    5.0
99986    5.0
99987    NaN
99988    5.0
99989    5.0
99990    5.0
99991    5.0
99992    2.0
99993    2.0
99994    2.0
99995    2.0
99996    2.0
99997    2.0
99998    2.0
99999    2.0
Name: Num_of_Loan, dtype: float64

Delay_from_due_date

In [63]:
df_clean["Delay_from_due_date"].head(20)

0      3
1     -1
2      3
3      5
4      6
5      8
6      3
7      3
8      3
9      7
10     3
11     3
12     3
13     3
14     3
15     3
16     5
17    13
18     8
19     8
Name: Delay_from_due_date, dtype: int64

In [64]:
df_clean["Delay_from_due_date"].describe()

count    100000.000000
mean         21.068780
std          14.860104
min          -5.000000
25%          10.000000
50%          18.000000
75%          28.000000
max          67.000000
Name: Delay_from_due_date, dtype: float64

In [65]:
df_clean["Delay_from_due_date"] = pd.to_numeric(
df_clean["Delay_from_due_date"],
errors="coerce")
df_clean.loc[
    df_clean["Delay_from_due_date"] < 0,
    "Delay_from_due_date"
] = np.nan

In [66]:
df_clean["Delay_from_due_date"].describe()

count    99409.000000
mean        21.207245
std         14.794693
min          0.000000
25%         10.000000
50%         18.000000
75%         28.000000
max         67.000000
Name: Delay_from_due_date, dtype: float64

Num_Credit_Inquiries

In [67]:
df_clean["Num_Credit_Inquiries"].head(20)

0     4.0
1     4.0
2     4.0
3     4.0
4     4.0
5     4.0
6     4.0
7     4.0
8     2.0
9     2.0
10    2.0
11    2.0
12    2.0
13    2.0
14    2.0
15    2.0
16    3.0
17    3.0
18    NaN
19    3.0
Name: Num_Credit_Inquiries, dtype: float64

In [68]:
df_clean["Num_Credit_Inquiries"].describe()

count    98035.000000
mean        27.754251
std        193.177339
min          0.000000
25%          3.000000
50%          6.000000
75%          9.000000
max       2597.000000
Name: Num_Credit_Inquiries, dtype: float64

In [69]:
df_clean["Num_Credit_Inquiries"] = pd.to_numeric(
df_clean["Num_Credit_Inquiries"],
errors="coerce")

Outstanding_Debt

In [70]:
df_clean["Outstanding_Debt"].tail(20)

99980     620.64
99981     620.64
99982     620.64
99983     620.64
99984    3571.7_
99985    3571.7_
99986     3571.7
99987     3571.7
99988     3571.7
99989     3571.7
99990     3571.7
99991     3571.7
99992     502.38
99993     502.38
99994     502.38
99995     502.38
99996     502.38
99997     502.38
99998     502.38
99999     502.38
Name: Outstanding_Debt, dtype: object

In [71]:
df_clean["Outstanding_Debt"].describe()

count      100000
unique      13178
top       1360.45
freq           24
Name: Outstanding_Debt, dtype: object

In [72]:
df_clean["Outstanding_Debt"] = (
df_clean["Outstanding_Debt"]
.astype(str)
.str.replace("_","",regex=False)
)
df_clean["Outstanding_Debt"] = pd.to_numeric(df_clean["Outstanding_Debt"], errors="coerce")

Credit_Utilization_Ratio

In [73]:
df_clean["Credit_Utilization_Ratio"].head(20)

0     26.822620
1     31.944960
2     28.609352
3     31.377862
4     24.797347
5     27.262259
6     22.537593
7     23.933795
8     24.464031
9     38.550848
10    33.224951
11    39.182656
12    34.977895
13    33.381010
14    31.131702
15    32.933856
16    28.616735
17    41.702573
18    26.519815
19    39.501648
Name: Credit_Utilization_Ratio, dtype: float64

In [74]:
df_clean["Credit_Utilization_Ratio"].describe()

count    100000.000000
mean         32.285173
std           5.116875
min          20.000000
25%          28.052567
50%          32.305784
75%          36.496663
max          50.000000
Name: Credit_Utilization_Ratio, dtype: float64

Total_EMI_per_month

In [75]:
df_clean["Total_EMI_per_month"].head(20)

0      49.574949
1      49.574949
2      49.574949
3      49.574949
4      49.574949
5      49.574949
6      49.574949
7      49.574949
8      18.816215
9      18.816215
10     18.816215
11     18.816215
12     18.816215
13     18.816215
14     18.816215
15     18.816215
16    246.992319
17    246.992319
18    246.992319
19    246.992319
Name: Total_EMI_per_month, dtype: float64

In [76]:
df_clean["Total_EMI_per_month"].describe()

count    100000.000000
mean       1403.118217
std        8306.041270
min           0.000000
25%          30.306660
50%          69.249473
75%         161.224249
max       82331.000000
Name: Total_EMI_per_month, dtype: float64

In [77]:
df_clean["Total_EMI_per_month"] = pd.to_numeric(
df_clean["Total_EMI_per_month"],
errors="coerce")

Amount_invested_monthly

In [78]:
df_clean["Amount_invested_monthly"].head(20)

0      80.41529543900253
1     118.28022162236736
2        81.699521264648
3      199.4580743910713
4     41.420153086217326
5     62.430172331195294
6      178.3440674122349
7     24.785216509052056
8       104.291825168246
9      40.39123782853101
10     58.51597569589465
11     99.30622796053305
12    130.11542024292334
13    43.477190144355745
14     70.10177420755677
15    218.90434353388733
16      168.413702679309
17    232.86038375993544
18             __10000__
19     825.2162699393922
Name: Amount_invested_monthly, dtype: object

In [79]:
df_clean["Amount_invested_monthly"].describe()

count         95521
unique        91049
top       __10000__
freq           4305
Name: Amount_invested_monthly, dtype: object

In [80]:
df_clean["Amount_invested_monthly"] = (
df_clean["Amount_invested_monthly"]
.astype(str)
.str.replace("_","",regex=False)
)

df_clean["Amount_invested_monthly"] = pd.to_numeric(
df_clean["Amount_invested_monthly"],
errors="coerce")

Monthly_Balance

In [81]:
df_clean["Monthly_Balance"].head(20)

0     312.49408867943663
1     284.62916249607184
2      331.2098628537912
3     223.45130972736786
4     341.48923103222177
5      340.4792117872438
6      244.5653167062043
7     358.12416760938714
8     470.69062692529184
9      484.5912142650067
10    466.46647639764313
11     465.6762241330048
12     444.8670318506144
13      481.505261949182
14     464.8806778859809
15    356.07810855965045
16    1043.3159778669492
17     998.8692967863226
18      715.741367403555
19     426.5134106068658
Name: Monthly_Balance, dtype: object

In [82]:
df_clean["Monthly_Balance"].describe()

count                                98800
unique                               98792
top       __-333333333333333333333333333__
freq                                     9
Name: Monthly_Balance, dtype: object

In [83]:
df_clean["Monthly_Balance"] = (
df_clean["Monthly_Balance"]
.astype(str)
.str.replace("_","",regex=False)
)

df_clean["Monthly_Balance"] = pd.to_numeric(
df_clean["Monthly_Balance"],
errors="coerce")

In [84]:
#Categorical columns: Check uniques values-Useful for spotting inconsistent labels.
cat_cols = df_clean.select_dtypes(include="object").columns
for col in cat_cols:
    print(col)
    print(df_clean[col].unique()[:10])

Month
['January' 'February' 'March' 'April' 'May' 'June' 'July' 'August']
Occupation
['Scientist' '_______' 'Teacher' 'Engineer' 'Entrepreneur' 'Developer'
 'Lawyer' 'Media_Manager' 'Doctor' 'Journalist']
Type_of_Loan
['Auto Loan, Credit-Builder Loan, Personal Loan, and Home Equity Loan'
 'Credit-Builder Loan' 'Auto Loan, Auto Loan, and Not Specified'
 'Not Specified' nan 'Credit-Builder Loan, and Mortgage Loan'
 'Not Specified, Auto Loan, and Student Loan'
 'Personal Loan, Debt Consolidation Loan, and Auto Loan'
 'Not Specified, and Payday Loan'
 'Credit-Builder Loan, Personal Loan, and Auto Loan']
Num_of_Delayed_Payment
['7' nan '4' '8_' '6' '1' '-1' '3_' '0' '8']
Changed_Credit_Limit
['11.27' '_' '6.27' '9.27' '5.42' '7.42' '6.42' '7.1' '11.1' '9.1']
Credit_Mix
['_' 'Good' 'Standard' 'Bad']
Credit_History_Age
['22 Years and 1 Months' nan '22 Years and 3 Months'
 '22 Years and 4 Months' '22 Years and 5 Months' '22 Years and 6 Months'
 '22 Years and 7 Months' '26 Years and 7 Months' '

Num_of_Delayed_Payment

In [85]:
df_clean["Num_of_Delayed_Payment"] = (
    df_clean["Num_of_Delayed_Payment"]
    .astype(str)
    .str.replace("_", "", regex=False)
)

df_clean["Num_of_Delayed_Payment"] = pd.to_numeric(
    df_clean["Num_of_Delayed_Payment"],
    errors="coerce"
)

df_clean.loc[
    df_clean["Num_of_Delayed_Payment"] < 0,
    "Num_of_Delayed_Payment"
] = np.nan

df_clean["Num_of_Delayed_Payment"].fillna(
    df_clean["Num_of_Delayed_Payment"].median(),
    inplace=True
)
#This is a numerical feature, and the median is robust to outliers.

C:\Users\a\AppData\Local\Temp\ipykernel_21152\3696422628.py:17: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_clean["Num_of_Delayed_Payment"].fillna(


Changed_Credit_Limit

In [86]:
df_clean["Changed_Credit_Limit"] = (
    df_clean["Changed_Credit_Limit"]
    .replace("_", np.nan)
)

df_clean["Changed_Credit_Limit"] = pd.to_numeric(
    df_clean["Changed_Credit_Limit"],
    errors="coerce"
)

df_clean["Changed_Credit_Limit"].fillna(
    df_clean["Changed_Credit_Limit"].median(),
    inplace=True
)

C:\Users\a\AppData\Local\Temp\ipykernel_21152\239737134.py:11: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_clean["Changed_Credit_Limit"].fillna(


Credit_History_Age

This column is not suitable as text for machine learning. Instead, convert it into total months.

In [87]:
#Create a conversion function
import re

def convert_credit_history(value):
    if pd.isna(value):
        return np.nan

    match = re.match(r'(\d+)\s+Years?\s+and\s+(\d+)\s+Months?', str(value))

    if match:
        years = int(match.group(1))
        months = int(match.group(2))
        return years * 12 + months

    return np.nan

In [88]:
#Apply function
df_clean["Credit_History_Age"] = (
    df_clean["Credit_History_Age"]
    .apply(convert_credit_history)
)

In [89]:
#Fill missing values
df_clean["Credit_History_Age"].fillna(
    df_clean["Credit_History_Age"].median(),
    inplace=True
)

C:\Users\a\AppData\Local\Temp\ipykernel_21152\1034665801.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_clean["Credit_History_Age"].fillna(


In [90]:
df_clean[[
    "Num_of_Delayed_Payment",
    "Changed_Credit_Limit",
    "Credit_History_Age"
]].head()

,Num_of_Delayed_Payment,Changed_Credit_Limit,Credit_History_Age
0,7.0,11.27,265.0
1,14.0,11.27,219.0
2,7.0,9.40,267.0
3,4.0,6.27,268.0
4,14.0,11.27,269.0


Occupation

In [91]:
df_clean["Occupation"] = (
df_clean["Occupation"]
.replace("_______",np.nan)
)
df_clean["Occupation"].fillna(
df_clean["Occupation"].mode()[0],
inplace=True)

#Fix inconsistent categories such as Occupation (Engineer, engineer, ENGINEER)
df_clean["Occupation"] = (
    df_clean["Occupation"]
    .str.strip()
    .str.title()
)

C:\Users\a\AppData\Local\Temp\ipykernel_21152\2872887542.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_clean["Occupation"].fillna(


In [92]:
df_clean["Occupation"].unique()

array(['Scientist', 'Lawyer', 'Teacher', 'Engineer', 'Entrepreneur',
       'Developer', 'Media_Manager', 'Doctor', 'Journalist', 'Manager',
       'Accountant', 'Musician', 'Mechanic', 'Writer', 'Architect'],
      dtype=object)

Credit_Mix

In [93]:
df_clean["Credit_Mix"] = (
df_clean["Credit_Mix"]
.replace("_",np.nan)
)

In [94]:
df_clean["Credit_Mix"].unique()

array([nan, 'Good', 'Standard', 'Bad'], dtype=object)

Payment_of_Min_Amount

In [95]:
df_clean["Payment_of_Min_Amount"] = (
df_clean["Payment_of_Min_Amount"]
.replace("NM",np.nan)
)

In [96]:
df_clean["Payment_of_Min_Amount"].unique()

array(['No', nan, 'Yes'], dtype=object)

Payment_Behaviour

In [97]:
df_clean["Payment_Behaviour"].unique()

array(['High_spent_Small_value_payments',
       'Low_spent_Large_value_payments',
       'Low_spent_Medium_value_payments',
       'Low_spent_Small_value_payments',
       'High_spent_Medium_value_payments', '!@9#%8',
       'High_spent_Large_value_payments'], dtype=object)

In [98]:
df_clean["Payment_Behaviour"] = df_clean["Payment_Behaviour"].replace("!@9#%8", np.nan)

In [99]:
print(df_clean["Payment_Behaviour"].value_counts(dropna=False))

Payment_Behaviour
Low_spent_Small_value_payments      25513
High_spent_Medium_value_payments    17540
Low_spent_Medium_value_payments     13861
High_spent_Large_value_payments     13721
High_spent_Small_value_payments     11340
Low_spent_Large_value_payments      10425
NaN                                  7600
Name: count, dtype: int64


In [100]:
df_clean[df_clean["Payment_Behaviour"].isna()].head()

,Month,Age,Occupation,Annual_Income,Monthly_Inhand_Salary,Num_Bank_Accounts,Num_Credit_Card,Interest_Rate,Num_of_Loan,Type_of_Loan,...,Credit_Mix,Outstanding_Debt,Credit_Utilization_Ratio,Credit_History_Age,Payment_of_Min_Amount,Total_EMI_per_month,Amount_invested_monthly,Payment_Behaviour,Monthly_Balance,Credit_Score
5,June,23.0,Scientist,19114.12,3093.745000,3.0,4.0,3,4.0,"Auto Loan, Credit-Builder Loan, Personal Loan,...",...,Good,809.98,27.262259,270.0,No,49.574949,62.430172,NaN,340.479212,Good
16,January,34.0,Lawyer,143162.64,12187.220000,1.0,5.0,8,3.0,"Auto Loan, Auto Loan, and Not Specified",...,Good,1303.01,28.616735,213.0,No,246.992319,168.413703,NaN,1043.315978,Good
32,January,21.0,Developer,35547.71,2853.309167,7.0,5.0,5,0.0,NaN,...,Standard,943.86,39.797764,368.0,Yes,0.000000,276.725394,NaN,288.605522,Standard
47,August,31.0,Lawyer,73928.46,5988.705000,4.0,5.0,8,0.0,NaN,...,Good,548.20,31.580990,390.0,No,0.000000,42.635590,NaN,796.234910,Good
54,July,34.0,Lawyer,10909427.00,3093.745000,0.0,1.0,8,2.0,"Credit-Builder Loan, and Mortgage Loan",...,Good,352.16,26.947565,373.0,No,911.220179,930.391898,NaN,326.241830,Good


Type_of_Loan

In [101]:
df_clean["Type_of_Loan"].unique()

array(['Auto Loan, Credit-Builder Loan, Personal Loan, and Home Equity Loan',
       'Credit-Builder Loan', 'Auto Loan, Auto Loan, and Not Specified',
       ..., 'Home Equity Loan, Auto Loan, Auto Loan, and Auto Loan',
       'Payday Loan, Student Loan, Mortgage Loan, and Not Specified',
       'Personal Loan, Auto Loan, Mortgage Loan, Student Loan, and Student Loan'],
      shape=(6261,), dtype=object)

This column contains multiple loan types in one cell. Don't split it yet. Keep it for feature engineering later, or encode it carefully depending on your modeling approach.

In [102]:
#Detect outliers using IQR
def outliers(col):

    Q1 = df_clean[col].quantile(0.25)
    Q3 = df_clean[col].quantile(0.75)

    IQR = Q3 - Q1

    lower = Q1 - 1.5*IQR
    upper = Q3 + 1.5*IQR

    return df_clean[(df_clean[col] < lower) | (df_clean[col] > upper)]

In [103]:
outliers("Annual_Income")

,Month,Age,Occupation,Annual_Income,Monthly_Inhand_Salary,Num_Bank_Accounts,Num_Credit_Card,Interest_Rate,Num_of_Loan,Type_of_Loan,...,Credit_Mix,Outstanding_Debt,Credit_Utilization_Ratio,Credit_History_Age,Payment_of_Min_Amount,Total_EMI_per_month,Amount_invested_monthly,Payment_Behaviour,Monthly_Balance,Credit_Score
54,July,34.0,Lawyer,10909427.0,3093.745000,0.0,1.0,8,2.0,"Credit-Builder Loan, and Mortgage Loan",...,Good,352.16,26.947565,373.0,No,911.220179,930.391898,NaN,326.241830,Good
231,August,40.0,Architect,6515990.0,2210.032500,6.0,3.0,15,3.0,"Payday Loan, Not Specified, and Auto Loan",...,Standard,400.07,23.177617,370.0,Yes,69.602920,51.830859,High_spent_Large_value_payments,339.569471,Standard
245,June,55.0,Scientist,586359.0,1684.168333,1.0,5.0,1,2.0,"Auto Loan, and Home Equity Loan",...,Good,413.32,28.836349,224.0,No,29.534584,68.956725,High_spent_Medium_value_payments,319.925525,Good
361,February,34.0,Entrepreneur,18334118.0,4369.482500,10.0,10.0,18,8.0,"Student Loan, Home Equity Loan, Payday Loan, D...",...,Bad,3422.49,36.729939,126.0,Yes,314.901785,124.541470,High_spent_Small_value_payments,257.504994,Standard
368,January,27.0,Architect,19717385.0,1057.435833,7.0,10.0,33,9.0,"Credit-Builder Loan, Personal Loan, Credit-Bui...",...,Bad,2797.17,25.314339,167.0,Yes,58.868441,60.550786,High_spent_Small_value_payments,246.324357,Standard
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
99664,January,31.0,Lawyer,5794377.0,3093.745000,5.0,4.0,9,1.0,Debt Consolidation Loan,...,Good,928.28,38.616320,219.0,No,72.250125,169.356592,High_spent_Large_value_payments,1210.848950,Standard
99714,March,26.0,Doctor,11712577.0,3093.745000,5.0,6.0,8,4.0,"Personal Loan, Payday Loan, Not Specified, and...",...,NaN,82.15,30.711678,277.0,No,113.833391,121.492500,Low_spent_Small_value_payments,334.657193,Standard
99721,February,53.0,Architect,8544730.0,3198.161667,5.0,4.0,5,3.0,"Home Equity Loan, Auto Loan, and Mortgage Loan",...,Good,1019.46,26.702382,200.0,No,86.809918,117.838682,High_spent_Medium_value_payments,365.167567,Standard
99882,March,31.0,Mechanic,16884797.0,5440.945000,7.0,3.0,1947,0.0,NaN,...,Standard,330.60,38.524095,276.0,Yes,0.000000,10000.000000,High_spent_Small_value_payments,536.342920,Standard


In [104]:
#Remove extra spaces
for col in df_clean.columns:
    if df_clean[col].dtype == "object":
        df_clean[col] = df_clean[col].str.strip()

In [105]:
#handle remaining null values

num_cols = df_clean.select_dtypes(include=np.number).columns

for col in num_cols:
    df_clean[col].fillna(df_clean[col].median(), inplace=True)

cat_cols = df_clean.select_dtypes(include="object").columns

for col in cat_cols:
    df_clean[col].fillna(df_clean[col].mode()[0], inplace=True)

C:\Users\a\AppData\Local\Temp\ipykernel_21152\1653985554.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_clean[col].fillna(df_clean[col].median(), inplace=True)
C:\Users\a\AppData\Local\Temp\ipykernel_21152\1653985554.py:11: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For 

In [106]:
#Final Missing value check: Should ideally return zeros or only intentional missing values.
df_clean.isnull().sum()

Month                       0
Age                         0
Occupation                  0
Annual_Income               0
Monthly_Inhand_Salary       0
Num_Bank_Accounts           0
Num_Credit_Card             0
Interest_Rate               0
Num_of_Loan                 0
Type_of_Loan                0
Delay_from_due_date         0
Num_of_Delayed_Payment      0
Changed_Credit_Limit        0
Num_Credit_Inquiries        0
Credit_Mix                  0
Outstanding_Debt            0
Credit_Utilization_Ratio    0
Credit_History_Age          0
Payment_of_Min_Amount       0
Total_EMI_per_month         0
Amount_invested_monthly     0
Payment_Behaviour           0
Monthly_Balance             0
Credit_Score                0
dtype: int64

In [107]:
#Save the cleaned dataset to a new CSV file
df_clean.to_csv(
    "../data/processed/cleaned.csv",
    index=False
)